src/ReactPlaceholder.tsx

```diff
@@ -1,128 +1,142 @@
-    const {
-      firstLaunchOnly, children, ready, className, // eslint-disable-line no-unused-vars
-      type, customPlaceholder, showLoadingAnimation, ...rest
-    } = this.props;
-
-    const classes = showLoadingAnimation ?
-      ['show-loading-animation', className].filter(c => c).join(' ') :
-      className;
+  customPlaceholder: React.ReactElement<{ [k: string]: any }> | null;
+  type?: undefined;
+  rows?: undefined;
+  color?: undefined;
+  showLoadingAnimation?: undefined;
+};
+
+type MediaPlaceholderProps = PlaceholderProps &
+  Omit<
+    React.ComponentProps<typeof placeholders.media>,
+    'color' | 'rows' | 'children'
+  > & {
+    type: 'media';
+  };
 
-    if (customPlaceholder && React.isValidElement(customPlaceholder)) {
-      const mergedCustomClasses = [
-        customPlaceholder.props.className,
-        classes
-      ].filter(c => c).join(' ');
-      return React.cloneElement(customPlaceholder, { className: mergedCustomClasses });
-    } else if (customPlaceholder) {
-      return customPlaceholder;
-    }
+type RectPlaceholderProps = PlaceholderProps &
+  Omit<React.ComponentProps<typeof placeholders.rect>, 'children'> & {
+    type: 'rect';
+  };
+
+type RoundPlaceholderProps = PlaceholderProps &
+  Omit<
+    React.ComponentProps<typeof placeholders.round>,
+    'color' | 'children'
+  > & {
+    type: 'round';
+  };
 
-    const Placeholder = placeholders[type!];
+type TextPlaceholderProps = PlaceholderProps &
+  Omit<
+    React.ComponentProps<typeof placeholders.text>,
+    'color' | 'rows' | 'children'
+  > & {
+    type: 'text';
+  };
 
-    return <Placeholder {...rest} className={classes} />;
+type TextRowPlaceholderProps = PlaceholderProps &
+  Omit<
+    React.ComponentProps<typeof placeholders.textRow>,
+    'color' | 'children'
+  > & {
+    type: 'textRow';
   };
 
-  setNotReady = () => {
-    const { delay } = this.props;
+export type Props =
+  | MediaPlaceholderProps
+  | RectPlaceholderProps
+  | RoundPlaceholderProps
+  | TextRowPlaceholderProps
+  | TextPlaceholderProps
+  | CustomPlaceholderProps;
+
+const ReactPlaceholder: React.FC<Props> = ({
+  delay = 0,
+  type = 'text',
+  color = '#CDCDCD',
+  rows = 3,
+  ready: readyProp,
+  firstLaunchOnly,
+  children,
+  className,
+  showLoadingAnimation,
+  customPlaceholder,
+  ...rest
+}) => {
+  const [ready, setReady] = React.useState(readyProp);
+  const timeout = React.useRef<null | number>(null);
+
+  const getFiller = (): React.ReactElement | null => {
+    const classes = showLoadingAnimation
+      ? joinClassNames('show-loading-animation', className)
+      : className;
 
-    if (delay && delay > 0) {
-      this.timeout = window.setTimeout(() => {
-        this.setState({ ready: false });
-      }, delay);
-    } else {
-      this.setState({ ready: false });
+    if (customPlaceholder && React.isValidElement(customPlaceholder)) {
+      const mergedCustomClasses = [customPlaceholder.props.className, classes]
```

Comment:
why not using `joinClassNames` here?

Reply:
Good question, not sure why I missed that either...